File: go/rerout/conversions.go
// Package rerout — Conversions namespace.
//
// Reached via Client.Conversions(). Wraps the /v1/conversions surface: record
// a conversion event against a click. Conversions attribute downstream value
// (purchases, sign-ups) back to the click that drove them.

package rerout

import "context"

// Conversions is the conversion-tracking namespace. Construct via
// Client.Conversions().
type Conversions struct {
	client *Client
}

// Record records a conversion event against a click via POST /v1/conversions.
//
// ClickID and EventName are required on input. The call is idempotent on the
// (click_id, event_name) pair: the returned ConversionResult reports whether a
// new row was Recorded and whether the event was a Duplicate of one already
// stored.
func (c *Conversions) Record(ctx context.Context, input RecordConversionInput) (*ConversionResult, error) {
	var out ConversionResult
	err := c.client.do(ctx, requestOptions{
		method: "POST",
		path:   "/v1/conversions",
		body:   input,
	}, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

File: rust/src/conversions.rs
//! `conversions` namespace — record conversion events against clicks.
//!
//! Conversions attribute downstream value (purchases, sign-ups) back to the
//! click that drove them, against `/v1/conversions`.

use crate::client::{HttpMethod, Rerout};
use crate::error::Result;
use crate::models::{ConversionResult, RecordConversionInput};

/// Conversion-tracking namespace. Reached via [`Rerout::conversions`].
#[derive(Debug, Clone)]
pub struct Conversions<'a> {
    client: &'a Rerout,
}

impl<'a> Conversions<'a> {
    pub(crate) fn new(client: &'a Rerout) -> Self {
        Self { client }
    }

    /// Record a conversion event against a click.
    ///
    /// `POST /v1/conversions` with [`RecordConversionInput`] as the JSON body.
    /// The call is idempotent on the `(click_id, event_name)` pair: the
    /// returned [`ConversionResult`] reports whether a new row was `recorded`
    /// and whether the event was a `duplicate` of one already stored.
    pub async fn record(&self, input: &RecordConversionInput) -> Result<ConversionResult> {
        self.client
            .request_json::<ConversionResult, _>(
                HttpMethod::Post,
                "/v1/conversions",
                None,
                Some(input),
            )
            .await
    }
}
